theano/tensor/basic.py

```diff
@@ -5910,7 +5921,9 @@ def __init__(self, offset=0, axis1=0, axis2=1):
 
     def make_node(self, x):
         x = as_tensor_variable(x)
-        assert x.ndim >= 2
+        if x.ndim < 2:
+            raise TypeError('Diagonal needs an input with 2 or more '
+                            'dimensions', x)
         return Apply(self, [x], [tensor(dtype=x.dtype,
                                         broadcastable=[False] * (x.ndim - 1))])
 
```

Comment:
You should copy the perform method from ExtractDiag so that it works on the GPU too.